pgcontents/query.py

```diff
@@ -510,7 +510,7 @@ def save_file(db, user_id, path, content, encrypt_func, max_size_bytes):
     )
     directory, name = split_api_filepath(path)
     with db.begin_nested() as savepoint:
-        try:
+        if not file_exists(db, user_id, path):
```

Comment:
I'm nervous about handling this by checking `file_exists` before executing the insert, since that allows for a possible race condition where another transaction inserts a file in between the `exists` check and the insert. That worry was the reason that this operation is currently implemented as "try to insert, but handle unique violations if they occur". Do you have any more info on your transaction that ended up in weird state?

Also, since this code was written, postgresql and sqlalchemy have added support for proper upserts directly in sql (see https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#insert-on-conflict-upsert, for instance). Using that would probably be the "right" way to fix this if there's a problem with the current error handling strategy.

Reply:
I'm new to SQLALchemy, so forgive my ignorance. 

I was under the assumption that the select and insert/update would happen in a single transaction, because of the `db.begin_nested()` in the with statement. Hence, if two transaction are trying to save (a new notebook) simultaneously, one would win, the other would need to retry a save and get updated in the second attempt. 

I get the following exceptions when using pgcontents version 0.6.0:

First, the insert fails
```
psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "uix_filepath_username"
DETAIL:  Key (user_id, parent_name, name)=(my_awesome_username, /, Untitled.ipynb) already exists.
```

Then 
```
sqlalchemy.exc.InvalidRequestError: Can't operate on closed transaction inside context manager.  Please complete the context manager before emitting further commands.
```

If I understand it correctly, that means that the transaction started by the `db.begin_nested()` has failed in some way and a new transaction needs to be started?

On the upsert: very interesting. My initial attempt at using the `on_conflict_do_update` fails with `Insert` does not have an attribute `on_conflict_do_update`. This is the code fragment:
```
res = db.execute(
            files.insert()
                .values(
                name=name,
                user_id=user_id,
                parent_name=directory,
                content=content,
            )
            .on_conflict_do_update(constraint="uix_filepath_username", set_={
                "name": name,
                "user_id": user_id,
                "parent_name": directory,
                "content": content
            })
        )
```